File: examples/clean_api_python/application/queries/get_products.py
"""GetProducts query + handler — Application / Queries"""
from dataclasses import dataclass, field
from typing import List, Dict, Any
import sys, os, time

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../../..')))

from backbone.domain.specifications import FilterParser, SortParser
from backbone.infrastructure.logging import LoggerFactory
from examples.clean_api_python.domain.entities.product import Product
from examples.clean_api_python.domain.repositories.product_repository import ProductRepository


@dataclass
class GetProductsQuery:
    """
    Read query for listing products.

    Query params:
      filters   — repeated: "field,operator,value[,condition]"
                  operators : eq ne gt gte lt lte contains in between is_null is_not_null
                  conditions: and (default) | or
      page      — page number (default 1)
      page_size — items per page (default 10)
      sort_by   — "field:direction"  e.g. "price:desc"  (default "created_at:desc")
    """
    filters: List[str] = field(default_factory=list)
    page: int = 1
    page_size: int = 10
    sort_by: str = "created_at:desc"


@dataclass
class GetProductsResult:
    products: List[Product]
    total_count: int
    page: int
    page_size: int


class GetProductsQueryHandler:

    def __init__(self, repository: ProductRepository):
        self.repository = repository
        self.logger = LoggerFactory.create_layer_logger(
            "product-api", "application", "GetProductsQueryHandler"
        )

    def handle(self, query: GetProductsQuery, context: Dict[str, Any]) -> GetProductsResult:
        self.logger.info("Handling GetProductsQuery", extra_data={
            "filters": query.filters, "page": query.page,
            "page_size": query.page_size, "sort_by": query.sort_by,
        })

        start = time.time()

        # Parse filters → composed Specification
        spec = FilterParser().parse_filters(query.filters) if query.filters else None

        # Parse sort_by → list of (field, direction) tuples
        sort_str = query.sort_by.replace(":", ",") if query.sort_by else "created_at,desc"
        sort_spec = SortParser().parse_sort(sort_str)
        sorts = sort_spec.to_sort_criteria() if sort_spec else [("created_at", "desc")]

        total_count = self.repository.count(spec)

        products = self.repository.find_by_criteria(
            filters=spec,
            sorts=sorts,
            page=query.page,
            page_size=query.page_size,
        )

        self.logger.info("GetProductsQuery handled", extra_data={
            "count": len(products), "total": total_count,
            "duration_ms": int((time.time() - start) * 1000),
        })

        return GetProductsResult(
            products=products,
            total_count=total_count,
            page=query.page,
            page_size=query.page_size,
        )

File: backbone-go/examples/clean-api-go/application/queries/get_products.go
// Package queries contains read-side application handlers (CQRS).
package queries

import (
	"context"
	"time"

	"github.com/freakjazz/backbone-go/application/exceptions"
	"github.com/freakjazz/backbone-go/domain/specifications"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/entities"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/repositories"
	"github.com/freakjazz/backbone-go/infrastructure/logging"
)

// GetProductsQuery is the read input for listing products.
//
// Query params (4 generic):
//
//	Filters  — repeated: "field,operator,value[,condition]"
//	           operators : eq ne gt gte lt lte contains in between is_null is_not_null
//	           conditions: and (default) | or
//	           examples  : "category,eq,Electronics,and"  "price,gt,500"
//	Page     — page number (default 1)
//	PageSize — items per page (default 10)
//	SortBy   — "field:direction"  e.g. "price:desc"  (default "created_at:desc")
type GetProductsQuery struct {
	Filters  []string
	Page     int
	PageSize int
	SortBy   string
}

// GetProductsResult is the read output.
type GetProductsResult struct {
	Products   []*entities.Product
	TotalCount int64
	Page       int
	PageSize   int
}

// GetProductsQueryHandler handles the GetProducts query.
type GetProductsQueryHandler struct {
	repository repositories.ProductRepository
	logger     *logging.EnhancedLogger
}

// NewGetProductsQueryHandler creates a new handler instance.
func NewGetProductsQueryHandler(repo repositories.ProductRepository, logger *logging.EnhancedLogger) *GetProductsQueryHandler {
	return &GetProductsQueryHandler{repository: repo, logger: logger}
}

// Handle executes the query.
func (h *GetProductsQueryHandler) Handle(ctx context.Context, query GetProductsQuery) (*GetProductsResult, error) {
	log := h.logger.
		WithLayer("application").
		WithHandler("GetProductsQueryHandler").
		WithMethod("Handle").
		WithContext(map[string]interface{}{
			"request_id":    ctx.Value("request_id"),
			"filters_count": len(query.Filters),
			"page":          query.Page,
			"page_size":     query.PageSize,
		})

	page := query.Page
	if page < 1 {
		page = 1
	}
	pageSize := query.PageSize
	if pageSize < 1 {
		pageSize = 10
	}
	if pageSize > 100 {
		pageSize = 100
	}

	sortField, sortDir := specifications.ParseSortBy(query.SortBy)
	criteria := specifications.ParseFilterParams(query.Filters, page, pageSize, sortField, sortDir)

	sql, args := criteria.GetFullSQL("SELECT * FROM products")
	log.Debug("Query criteria built", map[string]interface{}{"sql": sql, "args": args})

	start := time.Now()

	totalCount, err := h.repository.Count(ctx, criteria)
	if err != nil {
		log.ErrorWithCode("Failed to count products", 10002001, map[string]interface{}{"error": err.Error()})
		return nil, exceptions.NewUseCaseException("Failed to count products", "GetProductsQueryHandler")
	}

	products, err := h.repository.FindByCriteria(ctx, criteria)
	if err != nil {
		log.ErrorWithCode("Failed to fetch products", 10002002, map[string]interface{}{
			"error": err.Error(), "duration_ms": time.Since(start).Milliseconds(),
		})
		return nil, exceptions.NewUseCaseException("Failed to get products", "GetProductsQueryHandler")
	}

	log.Info("Products retrieved", map[string]interface{}{
		"count": len(products), "total": totalCount,
		"duration_ms": time.Since(start).Milliseconds(),
	})

	return &GetProductsResult{
		Products:   products,
		TotalCount: totalCount,
		Page:       page,
		PageSize:   pageSize,
	}, nil
}
